core/domain/exp_domain_test.py

```diff
@@ -4165,6 +4166,176 @@ def test_serialize_and_deserialize_returns_unchanged_exploration(
+        category = feconf.DEFAULT_EXPLORATION_CATEGORY
+        objective = feconf.DEFAULT_EXPLORATION_OBJECTIVE
+        language_code = constants.DEFAULT_LANGUAGE_CODE
+
+        content_id_generator = translation_domain.ContentIdGenerator()
+        init_state_dict = state_domain.State.create_default_state(
+            init_state_name,
+            content_id_generator.generate(
+                translation_domain.ContentType.CONTENT),
+            content_id_generator.generate(
+                translation_domain.ContentType.DEFAULT_OUTCOME),
+            is_initial_state=True).to_dict()
+
+        states_dict = {
+            init_state_name: init_state_dict
+        }
+
+        new_exploration = exp_domain.Exploration(
+              exploration_id, title, category, objective, language_code, [], '',
+              '', feconf.CURRENT_STATE_SCHEMA_VERSION - 1,
+              init_state_name, states_dict, {}, [], 0,
+              feconf.DEFAULT_AUTO_TTS_ENABLED,
+              content_id_generator.next_content_id_index, True)
+        sub_namespace = (
+          str(new_exploration.version) if new_exploration.version else None)
+        is_properly_cacheable = caching_services.set_multi(
+            namespace=caching_services.CACHE_NAMESPACE_EXPLORATION,
+            sub_namespace=sub_namespace,
+            id_value_mapping={exploration_id: new_exploration})
+
+        self.assertTrue(
+            is_properly_cacheable,
+            'Exploration is not properly cacheable'
+        )
+
+        new_exploration_retrieve_from_cache = caching_services.get_multi(
+            namespace=caching_services.CACHE_NAMESPACE_EXPLORATION,
+            sub_namespace=sub_namespace,
+            obj_ids=[exploration_id]
+        ).get(exploration_id)
+
+        self.assertNotEqual(
+            new_exploration_retrieve_from_cache,
+            None,
+            'Exploration is not properly cacheable'
+        )
+
+        default_exploration_dict = (
+            exp_domain.Exploration.create_default_exploration('exp_cache_test')
+            .to_dict()
+        )
+
+        exploration_data = (
+            new_exploration_retrieve_from_cache.to_dict()
+            if new_exploration_retrieve_from_cache
+            else default_exploration_dict)
+
+        updated_exploration = exp_domain.Exploration.migrate_state_schema(
+            exploration_data
+        )
+
+        new_exploration_dict = new_exploration.to_dict()
+
+        versioned_states = exp_domain.VersionedExplorationStatesDict(
+                states_schema_version=feconf.CURRENT_STATE_SCHEMA_VERSION - 1,
+                states=new_exploration_dict['states'])
+
+        conversion_fn = getattr(
+            exp_domain.Exploration, '_convert_states_v%s_dict_to_v%s_dict' % (
+            feconf.CURRENT_STATE_SCHEMA_VERSION - 1,
+            feconf.CURRENT_STATE_SCHEMA_VERSION))
+        migrated_exploration_dict = conversion_fn(
+                versioned_states['states'])
```

Comment:
I think this not the correct flow, right? Ideally the test should test the complete flow where you will be using `deserialize` method and in case the state-schema-version is not updated it should automatically be updated, right?

Reply:
Yes, you are right. Here flow of the programme is 
data -> serialize -> cache -> deserialize -> data
I put data in cache and retrieve it so state schema is updated.